### PR TITLE
feature(generic-list): support for pagination

### LINKED ISSUE #109: feat: show list of children using a ui recipe specified in the child
To show a list of objects a new generic UI plugin is needed. 

image

* Collapsible Object Table
- [x] Each object is shown as a row. 
- [x] There is a config element to define what items to show as rows for each child on the banner
- [x] The ui-plugin inside each element is defined based on the "View" concept as defined in the Grid plugin.
- [x] The table has general functions such as;
  - [x]  add
  - [x] delete
  - [x] move (reorder)
- [ ] (optional) Config allows for adding buttons, icon and call-back function, at the end of each row.
- [x] Default behaviour is to get the default view of each component. 
- [ ] Add pagination (view only) @soofstad 

* Object Table
Same as above, without the collapsible part. Instead the rows include editable data fields. 
- [x] There is a config element to define what items to to show on the columns
- [x] config has an option to be editable to read only. 
- [x] The table has general functions such as, add, delete, and move (reorder)
- [ ] (optional) Config allows for adding buttons, icon and call-back function, at the end of each row.
- [ ] Add pagination (view only)
 
Related issues:
- https://github.com/equinor/dm-core-packages/issues/127

### PR DESCRIPTION
**Pull request by soofstad**
## What does this pull request change?
- Adds pagination to "generic-list"-plugin

Screenshot_20230524_132650

## Issues related to this change
closes #109

### CODE REVIEW

File: packages/dm-core-plugins/src/generic-list/GenericListPlugin.tsx [hunk 1]
```diff
@@ -38,6 +45,22 @@ const Table = styled.table`
...
+): {
+  [key: string]: TGenericObject
+} {
+  const firstItem = currentPage * itemsPerPage
+  const lastItem = firstItem + itemsPerPage
+  const itemKeys = Object.keys(allItems).slice(firstItem, lastItem)
+  const newItems = Object.values(allItems).slice(firstItem, lastItem)
+  return Object.fromEntries(
+    itemKeys.map((key: string, index: number) => [key, newItems[index]])
+  )
```
Comment (ingeridhellen) [inline packages/dm-core-plugins/src/generic-list/GenericListPlugin.tsx:61]: You could instead do this, if you like:
```suggestion
  return Object.fromEntries(Object.entries(allItems).slice(firstItem, lastItem)) 
```

File: packages/dm-core-plugins/src/generic-list/GenericListPlugin.tsx [hunk 2]
```diff
@@ -161,98 +190,153 @@ export const GenericListPlugin = (
...
+          />
+        </div>
+        <div
+          style={{
+            display: 'flex',
+            flexDirection: 'row',
+            gridColumnStart: '3',
+            justifySelf: 'end',
+          }}
+        >
```
Comment (ingeridhellen) [inline packages/dm-core-plugins/src/generic-list/GenericListPlugin.tsx:332]: Could you extract this into a separate component in its own file?

File: packages/dm-core-plugins/src/generic-list/Components.tsx [hunk 3]
```diff
@@ -42,3 +42,16 @@ export const MoveItemDownButton = (props: { onClick: () => void }) => (
     <Icon data={arrow_down} title="Move down" />
   </Button>
 )
+
+export const ItemsPerPageButton = (props: {
+  onClick: (n: number) => void
+  setValue: number
```
Comment (ingeridhellen) [inline packages/dm-core-plugins/src/generic-list/Components.tsx:48]: Perhaps?:
```suggestion
  setItemsPerPage: (n: number) => void
  value: number
```
Usually in react, a variable whose name starts with "set" is a function.

**Review by ingeridhellen — APPROVED** [review]
Looks very nice! Some refactoring suggestions, though

File: packages/dm-core-plugins/src/generic-list/GenericListPlugin.tsx (hunk 1 above)
Comment (soofstad) [reply packages/dm-core-plugins/src/generic-list/GenericListPlugin.tsx:61]: Much better! Fixed :+1:

File: packages/dm-core-plugins/src/generic-list/GenericListPlugin.tsx (hunk 2 above)
Comment (soofstad) [reply packages/dm-core-plugins/src/generic-list/GenericListPlugin.tsx:332]: Replaced with @awesthouse suggestion

Comment (soofstad) [reply packages/dm-core-plugins/src/generic-list/GenericListPlugin.tsx:332]: Going to be a bit harder to get all the stuff aligned on one line. Also no that crucial  now that the "ItemsPerPage" buttons got more compact.

File: packages/dm-core-plugins/src/generic-list/Components.tsx (hunk 3 above)
Comment (soofstad) [reply packages/dm-core-plugins/src/generic-list/Components.tsx:48]: Renamed to "amount" :+1:

File: packages/dm-core-plugins/src/generic-list/GenericListPlugin.tsx [hunk 4]
```diff
@@ -161,98 +190,153 @@ export const GenericListPlugin = (
...
+        }}
+      >
+        <div
+          style={{
+            display: 'flex',
+            flexDirection: 'row',
+            alignItems: 'center',
+          }}
+        >
+          <Label label={'Items per page:'} />
```
[the remark this answers is not in the document]
Comment (soofstad) [reply packages/dm-core-plugins/src/generic-list/GenericListPlugin.tsx:295]: Thats how this should be done :man_facepalming: 
Thanks :smile:

**Pull request merged**